File: hash-map-set-table-(hashing)/781.py
# Tc = O(n)
# Sc = O(n)
from collections import defaultdict
import math
class Solution:
    def numRabbits(self, answers):
        mp = defaultdict(int)
        for x in answers:
            mp[x] += 1     
        total = 0
        for x, count in mp.items():
            group_size = x + 1
            groups = math.ceil(count / group_size)
            total += groups * group_size 
        return total

File: hash-map-set-table-(hashing)/781.js
//  Tc = O(n)
//  Sc = O(n)
function numRabbits(answers){
    const countMap = {}
    for(const x of answers){
        countMap[x] = (countMap[x] || 0) + 1
    }
    let total = 0
    for(const [x, count] of Object.entries(countMap)){
        const groupSize = parseInt(x) + 1
        const groups = Math.ceil(count / groupSize)
        total += groups * groupSize
    }
    return total
}